### ISSUE TITLE
Content_sets listed in modules are not searched when building image

### ISSUE THREAD

**Issue by kyguy** [cekit/cekit#757]
**Describe the bug**
The content_sets listed in modules are not searched when building our image. [1] We have an RPM, `kafka_exporter`, which is only available in the `amq-streams-2-for-rhel-8-rpms` content_set. We have `amq-streams-2-for-rhel-8-rpms` listed in our module's content_sets [2] but when running the following from our image descriptor directory [1]:
```
cekit build docker
```
the build fails because the  `kafka_exporter` RPM cannot be found:
```
2021-10-14 11:29:26,799 docker_builder.py:108        INFO  Docker: error: No package matches 'kafka_exporter'
2021-10-14 11:29:27,337 docker_builder.py:141        ERROR You can look inside the failed image by running 'docker run --rm -ti 285a370cd4ac bash'
...
```
We can get around this issue by listing the `amq-streams-2-for-rhel-8-rpms` content_set in content_sets section listed in our image descriptor file [3] but it makes no difference if it is listed in our module's content_set [2].

Is this intended behavor? Do all content_sets have to be listed in the content_sets of the image descriptor file even if they are already listed in a module?

**To reproduce**

From our image's parent directory [1] run:
```
cekit -v build docker
```

**Expected behavior**
We expect that the `amq-streams-2-for-rhel-8-rpms` content_set listed in the module's content_set [2] would be used to find/install our `kafka_exporter` RPM.

**Logs**
```
2021-10-14 11:29:26,799 docker_builder.py:108        INFO  Docker: error: No package matches 'kafka_exporter'
2021-10-14 11:29:27,337 docker_builder.py:141        ERROR You can look inside the failed image by running 'docker run --rm -ti 285a370cd4ac bash'
2021-10-14 11:29:27,337 cli.py:346        ERROR ('Image build failed, see logs above.', CekitError("Image build failed: 'The command '/bin/sh -c microdnf --setopt=install_weak_deps=0 --setopt=tsflags=nodocs install -y unzip tar rsync shadow-utils gzip kafka_exporter             && microdnf clean all             && rpm -q unzip tar rsync shadow-utils gzip kafka_exporter' returned a non-zero code: 1'"))
Traceback (most recent call last):
  File "/home/k/.local/lib/python3.9/site-packages/cekit/builders/docker_builder.py", line 89, in _build_with_docker
    raise CekitError("Image build failed: '{}'".format(error_message))
cekit.errors.CekitError: Image build failed: 'The command '/bin/sh -c microdnf --setopt=install_weak_deps=0 --setopt=tsflags=nodocs install -y unzip tar rsync shadow-utils gzip kafka_exporter             && microdnf clean all             && rpm -q unzip tar rsync shadow-utils gzip kafka_exporter' returned a non-zero code: 1'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/k/.local/lib/python3.9/site-packages/cekit/cli.py", line 339, in run
    command.execute()
  File "/home/k/.local/lib/python3.9/site-packages/cekit/builder.py", line 67, in execute
    self.run()
  File "/home/k/.local/lib/python3.9/site-packages/cekit/builders/docker_builder.py", line 242, in run
    image_id = self._build_with_docker(docker_client)
  File "/home/k/.local/lib/python3.9/site-packages/cekit/builders/docker_builder.py", line 150, in _build_with_docker
    raise CekitError(msg, ex)
cekit.errors.CekitError: ('Image build failed, see logs above.', CekitError("Image build failed: 'The command '/bin/sh -c microdnf --setopt=install_weak_deps=0 --setopt=tsflags=nodocs install -y unzip tar rsync shadow-utils gzip kafka_exporter             && microdnf clean all             && rpm -q unzip tar rsync shadow-utils gzip kafka_exporter' returned a non-zero code: 1'"))
```
**Environment information:**
 Operating system: Fedora 33
 Cekit version: 3.12.0

**Additional context**

[1] https://github.com/kyguy/amqstreams-1-openshift-image/tree/content_sets_reproducer/reproducer
[2] https://github.com/kyguy/amqstreams-1-openshift-image/blob/content_sets_reproducer/reproducer/modules/reproducer/module.yaml#L10
[3] https://github.com/kyguy/amqstreams-1-openshift-image/blob/content_sets_reproducer/reproducer/image.yaml#L21

(Edited to simplify reproducer)

**Comment by goldmann** [comment]
TL;DR; Content sets should be defined **on the image level**.

----

Hey Kyle!

Thanks for the report. I was investigating it today and I could reproduce the behavior you described in the report. You will probably not like the answer, but this is an expected behavior. Let me explain.

We are not parsing or validating the content of `content_sets` key (or external file content via `content_sets_file`). This is done on purpose because we don't want to be dependent on the potential schema changes of that file. What we get - we pass later, as-is, without any modification. Here is our schema definition for the `packages` section:

https://github.com/cekit/cekit/blob/ddb171c6ce1fad61bfbb52b4bca3c95a481d62b9/cekit/descriptor/packages.py#L15-L25

Look at lines 16-17. As you can imagine - without knowing the schema - we cannot make any sophisticated use of that content. This includes operation like merging. And this is what you basically requested in this report. In essence - **any `content_sets` directives defined in modules are ignored**.

You are correct that defining on the image level solves the problem. I wouldn't call it a workaround though. This is the correct place where it should be defined.

Sorry for the delay in answering this one!

Let me close this ticket, but feel free to reply!

**Comment by goldmann** [comment]
Additionally, to make things clear I'll make sure the content sets mentions in the module documentation will be removed and appropriate content will be added to image descriptor docs for content sets.

Reopening this ticket :)

**Comment by kyguy** [comment]
Understood, thanks for the information Marek!
